### PR TITLE
TLC calculation

### PR DESCRIPTION
**Pull request by janniclas**

### CODE REVIEW

File: adapter/src/main/kotlin/de/fraunhofer/iem/kpiCalculator/adapter/tools/tlc/TlcAdapter.kt [hunk 1]
```diff
@@ -10,17 +10,172 @@
...
+     * also considered.
+     *
+     * @param usedVersion version currently in use
+     * @param updateType do we try to update to the largest
+     * patch, minor or major version
+     * @param versions all available versions for the artifact
+     *
+     * @return highest applicable version
+     */
+    private fun getTargetVersion(
```
Comment (AnakinRaW) [inline adapter/src/main/kotlin/de/fraunhofer/iem/kpiCalculator/adapter/tools/tlc/TlcAdapter.kt:145]: make internal and test independently

**Review by AnakinRaW — CHANGES_REQUESTED** [review]

File: adapter/src/main/kotlin/de/fraunhofer/iem/kpiCalculator/adapter/tools/tlc/TlcAdapter.kt [hunk 2]
```diff
@@ -10,17 +10,172 @@
...
+
+    private fun getTechLagForNode(node: Node, artifacts: List<Artifact>, targetUpdateType: Version): TechLagResult {
+        val artifact = artifacts[node.artifactIdx]
+
+        val targetVersion = getTargetVersion(
+            usedVersion = node.usedVersion,
+            updateType = targetUpdateType,
+            versions = artifact.versions
+        )
+        val currentVersion = artifact.versions.find { it.versionNumber == node.usedVersion }
```
[the remark this answers is not in the document]
Comment (janniclas) [reply adapter/src/main/kotlin/de/fraunhofer/iem/kpiCalculator/adapter/tools/tlc/TlcAdapter.kt:117]: I understand the confusion. However, it's not necessarily the latest. target is depending on `targetUpdateType`, which tells us if we try to update to a major, minor or patch release. But `currentVersion` is what we found in the project, so maybe `usedVersion` would be a better name and also help?

File: adapter/src/main/kotlin/de/fraunhofer/iem/kpiCalculator/adapter/tools/tlc/TlcAdapter.kt [hunk 3]
```diff
@@ -10,17 +10,172 @@
...
+    private fun isProductionScope(scope: String, ecosystem: String): Boolean {
+        return when (ecosystem) {
+            "NPM" -> "dependencies" == scope
+            else -> true
+        }
+    }
+
+    private fun getTechLagForGraph(graph: Graph, artifacts: List<Artifact>, targetVersion: Version): TechLagResult {
+        if (graph.directDependencies.isEmpty()) {
+            return TechLagResult.Incomplete(
```
[the remark this answers is not in the document]
Comment (janniclas) [reply adapter/src/main/kotlin/de/fraunhofer/iem/kpiCalculator/adapter/tools/tlc/TlcAdapter.kt:91]: forgot to change the naming. Will change to `Empty`

File: adapter/src/main/kotlin/de/fraunhofer/iem/kpiCalculator/adapter/tools/tlc/model/Artifact.kt [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+    override fun compareTo(other: ArtifactVersion): Int = semver.compareTo(other.semver)
+
+    companion object {
+        fun create(versionNumber: String, releaseDate: Long, isDefault: Boolean = false): ArtifactVersion {
+            return ArtifactVersion(
+                releaseDate = releaseDate,
+                isDefault = isDefault,
+                // this step harmonizes possibly weired version formats like 2.4 or 5
+                // those are parsed to 2.4.0 and 5.0.0
+                versionNumber = validateAndHarmonizeVersionString(versionNumber)
```
[the remark this answers is not in the document]
Comment (janniclas) [reply adapter/src/main/kotlin/de/fraunhofer/iem/kpiCalculator/adapter/tools/tlc/model/Artifact.kt:40]: // Version parsing pattern: 1.2.3-alpha+build
    internal const val VERSION_REGEX: String = "^$CORE_VERSION$PRE_RELEASE?$BUILD?\$"

    // Prefixed version parsing pattern: v1.2-alpha+build
    internal const val LOOSE_VERSION_REGEX: String = "^v?$LOOSE_CORE_VERSION$PRE_RELEASE?$BUILD?\$"

the regex from the semver library we use (and the semver standard) doesn't allow that and will fail with an invalid version exception (which we should probably catch so we don't crash due to one library with weired format)

**Pull request merged**